### PR TITLE
#5 crud payments/transactions/statuses

### PR DESCRIPTION
**Pull request by lojzatran**
Part of https://github.com/commercetools/commercetools-paypal-plus-integration/issues/5

### CODE REVIEW

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java [hunk 1]
```diff
@@ -18,12 +24,33 @@
 public class PaymentServiceImpl extends BaseSphereService implements PaymentService {
 
     @Autowired
-    public PaymentServiceImpl(@Nonnull SphereClient sphereClient) {
+    public PaymentServiceImpl(SphereClient sphereClient) {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: why `@Nonnull` is removed?

File: src/main/java/com/commercetools/service/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+@Service
+public class OrderServiceImpl extends BaseSphereService implements OrderService {
+
+    @Autowired
+    public OrderServiceImpl(SphereClient sphereClient) {
+        super(sphereClient);
+    }
+
+    @Override
+    public CompletionStage<Optional<Order>> getByPaymentId(String paymentId) {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/service/impl/OrderServiceImpl.java]: paymentId is `@Nonnull` or `@Nullable`?

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java [hunk 3]
```diff
@@ -18,12 +24,33 @@
...
         super(sphereClient);
     }
 
     @Override
     public CompletionStage<Payment> getById(@Nullable String id) {
         return isBlank(id) ? completedFuture(null) : sphereClient.execute(PaymentByIdGet.of(id));
     }
+
+    @Override
+    public CompletionStage<Optional<Payment>> getByPaymentMethodAndInterfaceId(String paymentInterface, String interfaceId) {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: paymentInterface and interfaceId are `@Nonnull` or `@Nullable`?

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java [hunk 4]
```diff
@@ -18,12 +24,33 @@
...
+        }
+        PaymentQuery paymentQuery = PaymentQuery.of()
+                .withPredicates(p -> p.paymentMethodInfo().paymentInterface().is(paymentInterface))
+                .plusPredicates(p -> p.interfaceId().is(interfaceId));
+        return sphereClient.execute(paymentQuery)
+                .thenApply(PagedResult::head);
+    }
+
+    @Override
+    public CompletionStage<Payment> updatePayment(@Nonnull Payment payment, List<UpdateAction<Payment>> updateActions) {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: updateAction null or not?

File: src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void createOrderManuallyAndGetByPaymentId() {
+        PaymentDraftDsl paymentDraft = PaymentDraftBuilder.of(Money.of(22.33, EUR))
+                .build();
+        Payment ctPayment = executeBlocking(sphereClient.execute(PaymentCreateCommand.of(paymentDraft)));
+
+        createOrderWithPayment(ctPayment);
+
+        Optional<Order> orderOpt = executeBlocking(orderService.getByPaymentId(ctPayment.getId()));
+        assertThat(orderOpt.isPresent()).isTrue();
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: `.isPresent()).isTrue();` => `).isNotEmpty()`

File: src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java [hunk 6]
```diff
@@ -58,4 +68,42 @@ public void createManuallyAndGetWithWrongId() throws Exception {
...
+
+        executeBlocking(sphereClient.execute(PaymentCreateCommand.of(draft)));
+        Optional<Payment> payment = executeBlocking(
+                paymentService.getByPaymentMethodAndInterfaceId(paymentInterface, interfaceId)
+        );
+        assertThat(payment.isPresent()).isTrue();
+    }
+
+    @Test
+    public void shouldUpdateAmountPlanned() {
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java]: also add corner-case tests: with empty and null arguments

File: src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java [hunk 7]
```diff
@@ -58,4 +68,42 @@ public void createManuallyAndGetWithWrongId() throws Exception {
...
+        String paymentKey = "testPayment1";
+        Money amountBefore = Money.of(1, EUR);
+        PaymentDraftDsl draft = PaymentDraftBuilder.of(amountBefore)
+                .key(paymentKey)
+                .build();
+        Payment payment = executeBlocking(sphereClient.execute(PaymentCreateCommand.of(draft)));
+
+        Money amountAfter = Money.of(2, EUR);
+        List<UpdateAction<Payment>> updateActions = Collections.singletonList(ChangeAmountPlanned.of(amountAfter));
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java]: add test case with multiple update actions

File: src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java [hunk 8]
```diff
@@ -34,6 +38,20 @@ public static int cleanupPaymentTable(SphereClient sphereClient) {
         return removed;
     }
 
+    public static int cleanupOrders(SphereClient sphereClient) {
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java]: looks like it's time to make a common internal method, which makes logs and cleanup, to avoid copy-paste code every time

File: src/main/java/com/commercetools/service/impl/OrderServiceImpl.java (hunk 2 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/service/impl/OrderServiceImpl.java]: I put these annotations in the interface only.

Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/service/impl/OrderServiceImpl.java]: i saw them, but better to put on implementation as well, then (at least) IDE makes a warning if the implementation is different from the interface

File: src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java (hunk 5 above)
Comment (lojzatran) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: But there is no such method?

File: src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java (hunk 7 above)
Comment (lojzatran) [inline src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/e2986b9707006aab6c0fbaace6c820467011e093

File: src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java (hunk 6 above)
Comment (lojzatran) [inline src/test/java/com/commercetools/service/impl/PaymentServiceImplIntegrationTest.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/e2986b9707006aab6c0fbaace6c820467011e093

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java (hunk 1 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/1289fda27d11dd6414d251ad393158a2c15c001d

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java (hunk 3 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/1289fda27d11dd6414d251ad393158a2c15c001d

File: src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java (hunk 4 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/service/impl/PaymentServiceImpl.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/1289fda27d11dd6414d251ad393158a2c15c001d

File: src/main/java/com/commercetools/service/impl/OrderServiceImpl.java (hunk 2 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/service/impl/OrderServiceImpl.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/1289fda27d11dd6414d251ad393158a2c15c001d

**Comment by lojzatran** [comment]
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/e264f8bb1a19b9c3c926f706d2dc0eae0f92a735

File: src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java (hunk 5 above)
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: http://joel-costigliola.github.io/assertj/assertj-core-news.html#assertj-core-3.4.0-isNotEmpty

Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: ah, spring includes outdated _2.6.0_ version of AssertJ

Comment (lojzatran) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: Yes, but we don't have this version of assertJ, since we're using spring-boot-starter-test and it has only 2.6.0.: https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-test/1.5.4.RELEASE
I can upgrade to newest boot-start-test, WDYT? https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-test/2.0.0.M2

Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: ok, updated: 6a27d53 (merge from `dev`)

Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: > I can upgrade to newest boot-start-test, WDYT? https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-test/2.0.0.M2

No, this one is still not `RELEASE`, better to stick with stable versions. I've overridden the AssertJ, see the comment above

Comment (lojzatran) [inline src/test/java/com/commercetools/service/impl/OrderServiceImplIntegrationTest.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/21/commits/c8ee66faf063ffeeb9d633eb7fe878a9bd2c232a
